File: backend/src/main/java/com/pm/backend/model/Project.java
package com.pm.backend.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "projects")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Project {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotBlank
    @Column(nullable = false)
    private String status; // PLANNING, ACTIVE, COMPLETED, ARCHIVED

    @NotBlank
    @Column(nullable = false)
    private String priority; // LOW, MEDIUM, HIGH, CRITICAL

    private Double budget;
    @Builder.Default
    private Double spent = 0.0;
    
    private LocalDate deadline;
    @Builder.Default
    private Boolean isFavorite = false;
    private String colorLabel;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "project_members",
        joinColumns = @JoinColumn(name = "project_id"),
        inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    @Builder.Default
    private Set<User> members = new HashSet<>();

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: backend-node/src/models/Project.ts
import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../config/database';

export interface ProjectAttributes {
  id: number;
  name: string;
  description?: string;
  status: string;
  priority: string;
  budget?: number;
  spent?: number;
  deadline?: string;
  isFavorite?: boolean;
  colorLabel?: string;
  ownerId: number;
  createdAt?: Date;
  updatedAt?: Date;
}

export interface ProjectCreationAttributes extends Optional<ProjectAttributes, 'id'> {}

export class Project extends Model<ProjectAttributes, ProjectCreationAttributes> implements ProjectAttributes {
  declare id: number;
  declare name: string;
  declare description: string;
  declare status: string;
  declare priority: string;
  declare budget: number;
  declare spent: number;
  declare deadline: string;
  declare isFavorite: boolean;
  declare colorLabel: string;
  declare ownerId: number;

  declare readonly createdAt: Date;
  declare readonly updatedAt: Date;
}

Project.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    description: DataTypes.TEXT,
    status: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'PLANNING',
    },
    priority: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'MEDIUM',
    },
    budget: DataTypes.FLOAT,
    spent: {
      type: DataTypes.FLOAT,
      defaultValue: 0.0,
    },
    deadline: DataTypes.STRING,
    isFavorite: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },
    colorLabel: DataTypes.STRING,
    ownerId: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
  },
  {
    sequelize,
    tableName: 'projects',
  }
);
